src/Raven.Server/Documents/ShardedHandlers/ShardedBatchHandler.cs

```diff
@@ -97,6 +103,82 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
+        {
+            ClusterTransactionCommand.ValidateCommands(batch.ParsedCommands);
+
+            using (ServerStore.Cluster.ClusterTransactionWaiter.CreateTask(out var taskId))
+            {
+                var waitForIndexesTimeout = GetTimeSpanQueryString("waitForIndexesTimeout", required: false); //TODO  To check
+                var waitForIndexThrow = GetBoolValueQueryString("waitForIndexThrow", required: false) ?? true; //TODO  To check
+                var specifiedIndexesQueryString = HttpContext.Request.Query["waitForSpecificIndex"]; //TODO  To check
+
+                var disableAtomicDocumentWrites = GetBoolValueQueryString("disableAtomicDocumentWrites", required: false) ?? false;
+
+                BatchHandler.CheckBackwardCompatibility(HttpContext, ref disableAtomicDocumentWrites);
+
+                var options = new ClusterTransactionCommand.ClusterTransactionOptions(taskId, disableAtomicDocumentWrites, ClusterCommandsVersionManager.CurrentClusterMinimalVersion)
+                {
+                    WaitForIndexesTimeout = waitForIndexesTimeout,
+                    WaitForIndexThrow = waitForIndexThrow,
+                    SpecifiedIndexesQueryString = specifiedIndexesQueryString.Count > 0 ? specifiedIndexesQueryString.ToList() : null
+                };
+
+                var raftRequestId = GetRaftRequestIdFromQuery();
+                var clusterTransactionCommand = new ClusterTransactionCommand(
+                    ShardedContext.DatabaseName,
+                    ShardedContext.IdentitySeparator,
+                    batch.ParsedCommands,
+                    options,
+                    raftRequestId);
+
+                var clusterTransactionCommandResult = await ServerStore.SendToLeaderAsync(clusterTransactionCommand);
+                if (clusterTransactionCommandResult.Result is List<string> errors)
+                {
+                    HttpContext.Response.StatusCode = (int)HttpStatusCode.Conflict;
+                    throw new ConcurrencyException($"Failed to execute cluster transaction due to the following issues: {string.Join(Environment.NewLine, errors)}");
+                }
+                await ServerStore.WaitForCommitIndexChange(RachisConsensus.CommitIndexModification.GreaterOrEqual, clusterTransactionCommandResult.Index);
+                
+                DynamicJsonArray result;
+                if (clusterTransactionCommand.DatabaseCommands.Count > 0)
+                {
+                    using var timeout = new CancellationTokenSource(ServerStore.Engine.OperationTimeout);
+                    using var cts = CancellationTokenSource.CreateLinkedTokenSource(timeout.Token, HttpContext.RequestAborted);
+                    var databaseResult = await ServerStore.Cluster.ClusterTransactionWaiter.WaitForResults(taskId, cts.Token);
+                    result = databaseResult.Array;
```

Comment:
Isn't this a race condition? When we have many concurrent transactions and the buffer is cleared?

Reply:
Are you talking about `ClusterTransactionWaiter._results`?
Only the handler removes the result (by dispose the return value of `CreateTask`).